obspy/io/cmtsolution/core.py

```diff
@@ -0,0 +1,434 @@
+                         "contain a moment tensor element with an actual "
+                         "tensor.")
+    if not event.origins:
+        raise ValueError("Event must have at least one origin.")
+    if not event.magnitudes:
+        raise ValueError("Event must have at least one magnitude.")
+
+    # Attempt to get the body and surface wave magnitudes.
+    mb_candidates = \
+        [_i for _i in event.magnitudes if _i.magnitude_type == "Mb"]
+    ms_candidates = \
+        [_i for _i in event.magnitudes if _i.magnitude_type == "MS"]
+
+    if not mb_candidates:
+        warnings.warn("No body wave magnitude found. Will be replaced by the "
+                      "first magnitude in the event object.")
+        mb_mag = event.origins[0]
+    else:
+        mb_mag = mb_candidates[0]
+    if not ms_candidates:
+        warnings.warn("No surface wave magnitude found. Will be replaced by "
+                      "the first magnitude in the event object.")
+        ms_mag = event.origins[0]
+    else:
+        ms_mag = ms_candidates[0]
+
+    # Now find the cmt origin. First attempt to get the derived origin of
+    # the moment tensor,
+    if mt.derived_origin_id:
+        cmt_origin = mt.derived_origin_id.get_referred_object()
+    # Otherwise try to find the first one that is CMT
+    else:
+        candidates = [_i for _i in event.origins
+                      if _i.origin_type == "centroid"]
+        if candidates:
+            warnings.warn("No derived origin attached to the moment tensor. "
+                          "Will instead use another centroid origin to be "
+                          "written to the file.")
+            cmt_origin = candidates[0]
+        # Otherwise just take the preferred or first one.
+        else:
+            warnings.warn("Could not find a centroid origin. Will instead "
+                          "assume that the preferred or first origin is the "
+                          "centroid origin.")
+            cmt_origin = event.preferred_origin() or event.origins[0]
+
+    # Next step is to find a hypocentral origin.
+    candidates = [_i for _i in event.origins
+                  if _i.origin_type == "hypocenter"]
+    if candidates:
+        hypo_origin = candidates[0]
+    # Otherwise get the first one that is not equal to the CMT origin.
+    else:
+        if len(event.origins) == 1:
+            warnings.warn("Hypocentral origin will be identical to the "
+                          "centroid one.")
+            hypo_origin = event.origins[0]
+        else:
+            warnings.warn("No hypocentral origin could be found. Will choose "
+                          "the first one that is not identical to the "
+                          "centroid origin.")
+            hypo_origin = [_i for _i in event.origins if _i != cmt_origin][0]
+
+    # Try to find the half duration.
+    if mt.source_time_function:
+        if mt.source_time_function.duration:
```

Comment:
It actually different warnings depending on what happens so that is not possible here.